### ISSUE TITLE
Add editor command: Duplicate code

### ISSUE THREAD

**Issue by justcla** [justcla/HotCommands#4]
Suggested key binding: Ctrl+D

**Comment by justcla** [comment]
Aakash will look into this.

**Comment by pramod-hegde** [comment]
need to handle block selection after paste

**Comment by aakash-1906** [comment]
Resolved the issue. Pls refer the screenshots in attachment of working version.!
Works a little better than Resharper. And that too for multi-line vertical select and for Forward/Reverse select too.!
Resharper doesn't do it properly for reverse select OR multi-line vertical block select.!
Let me know your comments.!
And sorry if you aren't able to view attachment. In that case pls ping me


[Select Words inside single line.docx](https://github.com/justcla/HotCommands/files/372975/Select.Words.inside.single.line.docx)

**Comment by justcla** [comment]
Nice work Aakash.
I think multi-line select duplicate should apply single line duplicate to each line - and not apply nice line breaks (as in the documented example #3).

**Comment by aakash-1906** [comment]
Done with as you suggested and checked in

**Comment by aakash-1906** [comment]
Done with both Ctrl+D and Ctrl+Shift+D

**Comment by carsonal** [comment]
Is this closed?

**Comment by aakash-1906** [comment]
Justin should make a call. Its working fine for Text editors. Not for XML based files though. So, I feel we may close this now and open a new one for XML (or) just leave this open if fixing for XML is something very simple and if anyone could help with some pointers on it.

**Comment by verysimplenick** [comment]
Duplicate line works not so perfectly (as Resharper for example), I don't want get duplicated line in copy-paste buffer.

**Comment by justcla** [comment]
Hey Nick. I agree about it not adding to the copy-paste buffer. In fact, I would say, even if there is no selection, it should still not add to the buffer.
Do you think you could pull down the source and code that in?

**Comment by aakash-1906** [comment]
Ha well Justin. It getting added to copy-paste buffer is something we didn't think about earlier. I feel, we will be able to remove the recent item from that copy-paste buffer, and that shall resolve it.

However, this is better than Resharper in some ways. For it supports block (horizontal block) selection , in addition to multi-line select. And works perfectly for forward and reverse selection too.!

**Comment by justcla** [comment]
I think if we use:

editorOperations.InsertText(selectedText);

instead of

                editorOperations.CopySelection();
                editorOperations.MoveToNextCharacter(false);
                editorOperations.Paste();
                editorOperations.MoveToPreviousCharacter(false);

that should do the trick.

Not sure about moving the position of the caret, though. I was trying to test it now, but I notice it is referencing Dev15 assemblies because I upgraded it to VS2017.
Might need to wind that back to Dev14 so it works on both VS2015 and VS2017.

**Comment by justcla** [comment]
The feature works really well, I must admit.

Although, the code could probably be cleaned up a bit.
For instance, there is no need for the variable "singleLine". Those two if blocks can be merged.
Also, there seems to be no need for "text". It is just a copy of "selectedText".
And finally, I wonder if there is a lot of duplicate of code in those blocks that deal with single selection or multi-line selection. Perhaps a good chunk of the "code and paste and move next" work could be moved to a method and called from a few places to reduce redundant code.
Those are my thoughts.

**Comment by yannduran** [comment]
This works perfectly for me so far 👍

**Comment by verysimplenick** [comment]
> This works perfectly for me so far
@yannduran, Yes it's works, but replace copy-paste buffer if you not select line. I'm really sad about this :(

**Comment by justcla** [comment]
@verysimplenick I can release a fix as soon as you code it up for me! :-D
This is a community project. We value your input and contributions.
If you can submit a pull request, I'll test it, accept it and release it pronto.
Thanks

**Comment by verysimplenick** [comment]
@justcla , I don't have addin sdk for now.

**Comment by justcla** [comment]
Once we've fixed the caret position and the Last-Line issue, I will merge it into Master.

**Comment by justcla** [comment]
I've fixed the caret position issue. And I've included a hack to get around the issue of the last line not having a newline character.
Branch marged to Master.
New release created: v1.1.4
and now available on Marketplace.
:-)
